torch/utils/data/datasets/groupbyfilenamedataset.py

```diff
@@ -0,0 +1,119 @@
+from torch.utils.data.dataset import IterableDataset
+from typing import List, Iterable, Iterator
+from collections import OrderedDict
+
+import os
+
+class GroupByFilenameIterableDataset(IterableDataset):
+    r""" :class:`GroupByFilenameIterableDataset`.
+
+    IterableDataset to group binary streams from input iterables by pathname without extension,
+    yield a list with `group_size` items in it, each item in the list is a tuple of
+    pathname and binary stream
+
+    args:
+        dataset: Iterable dataset that provides pathname and zip binary stream in tuples
+        group_size: the size of group
+        buffer_size: the size of buffer which is used to store non-grouped but iterated streams
+        length: a nominal length of the dataset
+    """
+
+    def __init__(
+            self,
+            dataset : Iterable,
+            group_size : int = 1,
+            buffer_size : int = 10,
+            length: int = -1):
+        super().__init__()
+        self.dataset : Iterable = dataset
+        self.group_size : int = group_size
+        self.buffer_size : int = buffer_size
+        self.stream_buffer : OrderedDict = OrderedDict()
+        self.length : int = length
+
+
+    def __scan_stream_buffer(self, group_size : int):
+        # scan current stream buffer, return a group of data or partial group of data
+        stream_buffer = self.stream_buffer
+        res = []
+        removed_pathnames = []
+        key = None
+        for pathname, stream in stream_buffer.items():
+            name_ext = os.path.splitext(pathname)
+            if key is None:
+                key = name_ext[0]
+                res.append((pathname, stream))
+                removed_pathnames.append(pathname)
+            elif key == name_ext[0]:
+                res.append((pathname, stream))
+                removed_pathnames.append(pathname)
+            if len(res) == group_size:
+                break
+        for pathname in removed_pathnames:
+            del stream_buffer[pathname]
+        return res
```

Comment:
This strategy requires to find a full group with the key of the first item within the buffer.

For example of group_size 2, if the current buffer holds [a1, b1, c1, d1, d2], and next items are c2, b2, a2. It will blow up the buffer even if there is a group existing in the buffer.

IMO, it's not safe or reasonable to always raise error for users and require them change the order of dataset.

Reply:
Besides, it's even worse when we enable the multiprocess or multithread. We loss control of the order of datasets.

I'm not sure it's reasonable, but I propose to have some mechanisms to depreciate items existing in the buffers for too long time.

Reply:
@ejguan 
1. For buffer blow up problem. Yes, it will blow up if file streams are not well ordered. And the reason of having this linear scan algo is that, we need to maintain the order of the grouped item in the following way: if we see a1 first, we need to return (a1, a2) before (b1, b2) even if b2 is before a2, since a1 is the first item discovered, we don't want to jump any items. If the data set has data problem, say disordered too much, or missing items, then just raise error to user. 

2. I think the order of dataset won't be affected even under multiprocess or multithread. The order of data is always first seen first serve (given we don't jump/bypass any items).  Dis-ordered data is something user should take care of, Nvidia's dataset even put a strict rule that grouped data must be sit next to each other in the input dataset, finding data back and force will slow down the grouping process. 

We don't have much error tolerance here, I discussed this with @VitalyFedyunin a bit earlier together with other datasets (eg, extract tar dataset) as well, I think we can discuss more for grouping dataset.